### PR TITLE
Add bpmn:InclusiveGateway support

### LINKED ISSUE #88: Inclusive Gateways support
Hello!
Is it planned to implement support for inclusive gateways?
image

### PR DESCRIPTION
**Pull request by barmac**
https://user-images.githubusercontent.com/28307541/152647029-07e03495-ce98-45b4-9998-45e41db54866.mov

- [x] Outgoing sequence flow enable/disable
- [x] Default flow handling
- [x] Sync join

Closes #88

### CODE REVIEW

**Comment by barmac** [comment]
https://user-images.githubusercontent.com/28307541/152649822-0ce9724f-bc85-4de7-8aa6-e303bfeb1fbc.mov

I've added default flow handling. You cannot select it manually but it will be selected when you disable remaining flows.

**Comment by barmac** [comment]
Currently, we don't allow to synchronize the tokens. As a result, a converging inclusive gateway is equal to a converging exclusive gateway, i.e. tokens of the same execution don't wait one for another. 

I checked how Camunda Platform handles the synchronization (which is enforced) and it seems to count tokens waiting at the gateway and traverse the graph in order to count possible paths to the gateway. If there are no more tokens which can reach the gateway, the execution is resumed. Cf. https://github.com/camunda/camunda-bpm-platform/blob/master/engine/src/main/java/org/camunda/bpm/engine/impl/bpmn/behavior/InclusiveGatewayActivityBehavior.java#L141

If I were to implement it in a similar way, I need to find out how to access the currently active tokens within the execution. Any hints will be appreciated.

**Comment by barmac** [comment]
https://user-images.githubusercontent.com/28307541/152663448-1a204ba9-cf5a-438d-98e3-dd7742e0b1fd.mov

Now we enforce synchronization on join. It waits for any tokens which can possibly reach the inclusive gateway. 

Via https://github.com/bpmn-io/bpmn-js-token-simulation/pull/110/commits/a30d5e76738fa28d2de33eb8bde2b5f883da6521 I added Camunda Platform-like behavior that when there are enough tokens waiting at the gateway as the number of incoming flows. I removed it afterwards, because this  leads sometimes to a deadlock when there is a fake-join or a parallel gateway is not joined. Not sure if we should account for bad practices. Still, I'd rather avoid deadlock in the simulation when it's possible.

Is it how it's defined in the spec? There's no real specification on the inclusive gateway. Therefore, I can't determine the correct behavior from the PDF.

I will add a couple of test cases later and then I think we can initiate the review.

**Comment by barmac** [comment]
I've added the following changes:

* Join is now handled as in Camunda Platform, i.e. we wait for at most as many tokens as there are incoming sequence flows; this is similar to how we handle the parallel gateway. If there are no more tokens which have a direct route to the gateway, the join is performed with smaller number of tokens.
* For fork, when user deactivates the last active sequence flow and there is no default flow, another flow is activated. This is to prevent an error when no outgoing flows are active. We could roll this back if there is a way to notify incidents in the simulation.

There is one problem for which we'd need to add more changes to the library. The switch to activate/deactivate a sequence flow is attached to the connection. Then, the context pads module does not allow to set the position of the overlay. As a result, when there are multiple sequence flows starting from the same point, the switches overlap. Have a look at the screenshot:

image

**Comment by barmac** [comment]
There is a bug when running multiple simulations at once. We don't properly check the parent context.

Also, we should avoid a non-spec-compliant deadlock when a boundary event is triggered after other tokens arrived.

**Comment by nikku** [comment]
Moving this back to backlog. Let us await further feedback before actually merging a :magic_wand: :tada: :space_invader: element behavior.

**Comment by barmac** [comment]
Sure, let's wait for Gandalf the White to figure this out ^^

**Comment by JaySmith** [comment]
How can I get this functionality?  I would love to have this.

**Comment by nikku** [comment]
@JaySmith Please refer to https://github.com/bpmn-io/bpmn-js-token-simulation/issues/88#issuecomment-961841600. 

What is your practical use of inclusive gateway support within this tool?

**Comment by barmac** [comment]
> How can I get this functionality? I would love to have this.

Feel free to fix the remaining issues of this implementation: https://github.com/bpmn-io/bpmn-js-token-simulation/pull/110#issuecomment-1047767335

**Comment by JaySmith** [comment]
> @JaySmith Please refer to [#88 (comment)](https://github.com/bpmn-io/bpmn-js-token-simulation/issues/88#issuecomment-961841600).
> 
> What is your practical use of inclusive gateway support within this tool?

I am documenting our project lifecycle process and at the end our project managers need to perform several activities.  For example if they have a open purchase order they need to close it, if they have a capital expenditure request they need to close that as well, if they have implemented a new software solution that needs budget for future maintenance and licensing they need to have that added to the next years budget.

Not all of these will apply.  Is there a different way to do that?

Jay

**Comment by JaySmith** [comment]
> > How can I get this functionality? I would love to have this.
> 
> Feel free to fix the remaining issues of this implementation: [#110 (comment)](https://github.com/bpmn-io/bpmn-js-token-simulation/pull/110#issuecomment-1047767335)

If I could code this I would love to contribute.

**Comment by barmac** [comment]
I've fixed the remaining issues. I believe this is now BPMN-compliant.

https://github.com/bpmn-io/bpmn-js-token-simulation/assets/28307541/c903cbda-5a8d-44b4-b21a-cf38911e7de3

File: lib/simulator/behaviors/InclusiveGatewayBehavior.js [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+/**
+ * Returns true if there are either no remaining scopes in the parent scope, or if they are not
+ * reachable from the current element.
+ *
+ * @returns {Boolean}
+ */
+InclusiveGatewayBehavior.prototype._canJoin = function(context) {
+  const remainingScopes = this._getRemainingScopes(context);
+
+  // There are still some tokens to wait for.
```
Comment (philippfromme) [inline lib/simulator/behaviors/InclusiveGatewayBehavior.js:78]:
```suggestion
  // there are still some tokens to wait for
```

**Comment by nikku** [comment]
@barmac This looks great! :star_struck: 

Will give it a comprehensive review tomorrow.

**Review by nikku — COMMENTED** [review]
Thanks for your contribution. It is great:

* It uses the batteries included facilities the engine already offers (waiting for scopes changing and more :heart:)
* It is really simple at the core
* It even supports default flows in a convenient manner :tada: 

----

Please check a bunch of improvements I'd put on top, namely a simplified (and unified) join behavior, and re-using the `activityBehavior` to advance execution after a join + a bunch of extra tests.

**Comment by barmac** [comment]
Thanks for the review and the improvements to the PR. I think the changes make sense.

There is still a CI job failing and it complains about CSS file not being present: https://github.com/bpmn-io/bpmn-js-token-simulation/actions/runs/7038213433/job/19154806002?pr=110 I will check if we can fix it easily.

**Comment by nikku** [comment]
> https://github.com/bpmn-io/bpmn-js-token-simulation/pull/110#issuecomment-1833312837

This we can ignore, as it is a non-lethal change. In fact the actual test failure is due to a shaky test case. Unfortunately I was not able to further stabilize the test cases as things work just fine locally for me.

**Comment by nikku** [comment]
Squashed the changes on this branch (where it makes sense).

**Comment by barmac** [comment]
I can reproduce the CSS issue but not the test failing. Do we really need to support bpmn-js@8 though? @nikku

**Comment by barmac** [comment]
Sorry, I missed your comments above. Still, I think we could safely move to bpmn-js@11 or even higher at no cost.

**Comment by nikku** [comment]
> Sorry, I missed your comments above. Still, I think we could safely move to bpmn-js@11 or even higher at no cost.

We're testing against both legacy and new versions, and I think that is fine. Or do you propose to ditch our legacy tests? We'll want to do that at some point, yes.

**Comment by barmac** [comment]
Given that the legacy test fails for no meaningful reason, the failure is not reproducible locally, and the version the test runs against is 2 years old and 7 major versions behind, my suggestion is to leave v8 behind and test against a version closer to the current one. Or we find out why the test fails.

**Comment by nikku** [comment]
@barmac The test failure is non-deterministic, and can happen on the current version of bpmn-js, too.

**Comment by nikku** [comment]
Cf. https://github.com/bpmn-io/bpmn-js-token-simulation/actions/runs/7048800536/job/19185774908?pr=110 :see_no_evil:

**Comment by nikku** [comment]
Merged, as all test failures are unrelated to this PR.

**Comment by barmac** [comment]
Oh, that's terrible :(

**Pull request merged**